source/agora/common/EnrollmentManager.d

```diff
@@ -0,0 +1,407 @@
+        {
+            log.info("Rejected enroll: {}", enroll);
+            return false;
+        }
+
+        () @trusted { assumeSafeAppend(buffer); } ();
+
+        scope SerializeDg dg = (scope const(ubyte[]) data) nothrow @safe
+        {
+            buffer ~= data;
+        };
+
+        serializePart(enroll, dg);
+
+        () @trusted {
+            db.execute("INSERT INTO validator_set (key, val) VALUES (?, ?)",
+                enroll.utxo_key[], buffer);
+        }();
+
+        return true;
+    }
+
+    /***************************************************************************
+
+        Returns:
+            the number of validators in the validator set
+
+    ***************************************************************************/
+
+    public size_t getEnrollmentLength () @safe
+    {
+        return () @trusted {
+            return db.execute("SELECT count(*) FROM validator_set").oneValue!size_t;
+        }();
+    }
+
+    /***************************************************************************
+
+        Walk over the enrollment data in the set and call the provided delegate
+        with each hash and enrollment data
+
+        Params:
+            dg = the delegate to call
+
+        Returns:
+            the status code of the delegate, or zero
+
+    ***************************************************************************/
+
+    public int opApply (scope int delegate(const ref Hash hash, 
+        const ref Enrollment enroll) dg) @trusted
+    {
+        auto results = this.db.execute("SELECT key, val FROM validator_set");
+
+        foreach (row; results)
+        {
+            auto hash = *cast(Hash*)row.peek!(ubyte[])(0).ptr;
+            auto enroll = deserializeFull!Enrollment(row.peek!(ubyte[])(1));
+
+            if (auto ret = dg(hash, enroll))
+                return ret;
+        }
+
+        return 0;
+    }
+
+    /***************************************************************************
+
+        Remove the enrollment data with the given key from the validator set
+
+        Params:
+            enroll_hash = key for an enrollment data to remove
+
+    ***************************************************************************/
+
+    public void removeEnrollment (const ref Hash enroll_hash) @trusted
+    {
+        this.db.execute("DELETE FROM validator_set WHERE key = ?", enroll_hash[]);
+    }
+
+    /***************************************************************************
+
+        Make an enrollment data for enrollment process
+
+        Params:
+            frozen_utxo_hash = the hash of a frozen UTXO used to identify 
+                                a validator and to generate a siging key
+
+        Returns:
+            an enrollment data
+
+    ***************************************************************************/
+
+    public Enrollment createEnrollment (const ref Hash frozen_utxo_hash) @trusted
+    {
+        this.data = new Enrollment();
```

Comment:
Why is it a pointer ?

Reply:
Also is this function supposed to be called multiple times ? In what order ? For what purpose ?
The documentation is very generic. The fact that it is stored as a member of this class tells me that either:
- There can only be one
OR
- We cache the last one

If there can only be one, it feels like it shouldn't be called multiple times. If we cache the last created Enrollment data, why do we do that ? Neither the documentation of this function nor the variable says anything.

Reply:
@Geod24 I'd like to discuss with you about this. I wanna split the EnrollmentManager into the ValidatorSet which manage other validator's information and the EnrollmentManager which manage the enrollment data for a node itself.